[MLIR][XLA] Buffer Assignment

ehsan-nt opened the pull request:
In this PR, we have provided our first draft for generic XLA buffer assignment. This generic BufferAssignment class automatically analyzes the values and their aliases (also in other blocks) and returns the positions of allocations and deallocations. To find these positions, the algorithm uses the block Dominator and Post-Dominator analyses. In our proposed algorithm, we have considered aliasing, liveness analysis, nested regions, branches, conditional branches, critical edges, and independency to custom block terminators. This implementation doesn't support block loops. However, we have considered this in our design. For this purpose, we only need to have a loop analysis to allocate and deallocate some special cases outside of these loops. 

This is the sample usage of BufferAssignment class:
```c++
BufferAssignment assignments(funcOp);
auto positions = assignments.computeAllocAndDeallocPositions(value);
allocBuilder.setInsertionPoint(positions.getAllocPosition());
<create alloc>
deallocBuilder.setInsertionPointAfter(positions.getDeallocPosition());
<create dealloc>
```
Or alternatively:
```c++
BufferAssignment assignments(funcOp);
auto positions = assignments.computeAllocAndDeallocPositions(value);
positions.insertAlloc<AllocOp>(...);
positions.insertDealloc<DeallocOp>(...);
```

Please note that this class should be used during the legalization (i.e. HLO-To-LHLO).

dfki-mako committed:
Added new buffer_assignment module to build system.

Commit:
Added new straight-forward alias analysis to determine all potential aliases of a single value.

dfki-mako committed:
Added specialized dominator analysis and initial BufferAssignment interface.

dfki-mako committed:
Implemented findPlacementBlock to get the actual block for placing allocs and deallocs.

Commit:
Implemented missing getAllocPosition and getDeallocPosition.

dfki-mako committed:
Added utility methods to insert allocs and deallocs.

Commit:
Added dump methods to BufferAssignment class.

dfki-mako committed:
Added additional documentation about the algorithm, the class usage and the general limiations.

Commit:
Added test cases for new BufferAssignment analysis.

pifon2a reviewed (commented):
First portion of nits. More coming.

pifon2a commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h:
```diff
@@ -0,0 +1,185 @@
+  /// Initializes the internal mappings.
+  void build(MutableArrayRef<Region> regions);
+
+ private:
```
not need to write `private:` twice

pifon2a commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h:
```diff
@@ -0,0 +1,185 @@
+/// deallocs.
+class BufferAssignmentAliasAnalysis {
+ public:
+  using ValueSetT = SmallPtrSet<Value, 16>;
```
Will buffer_assignment.* live in tensorflow repo or will it be moved to llvm/mlir? Asking because of the style. It looks pretty prepared to be moved to llvm codebase.

dfki-mako replied:
Yes and no :) Currently, it would be better to continue the development of the BufferAssignment class in tensorflow since it is particularly designed for XLA/LHLO.

pifon2a commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h:
```diff
@@ -0,0 +1,185 @@
+
+namespace mlir {
+namespace xla {
+
```
nit: no need for the empty line between namespaces. But maybe add one after `namespace detail {` to separate the comment.

pifon2a commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,281 @@
+/// all newly found potential aliases in the given result set.
+void BufferAssignmentAliasAnalysis::resolveRecursive(Value value,
+                                                     ValueSetT& result) const {
+  if (!result.insert(value).second) return;
```
i think in TF codebase this should be in {}. Since this code already follows LLVM style, why not putting it to LLVM repo?

pifon2a commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,281 @@
+/// that will be passed to the corresponding block arguments and inserts them
+/// into map.
+void BufferAssignmentAliasAnalysis::build(MutableArrayRef<Region> regions) {
+  for (Region& region : regions)
```
{} with this `for`?

pifon2a commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,281 @@
+      for (auto pred : block.getPredecessors()) {
+        // Determine the current successor index of the current predecessor.
+        unsigned successorIndex = 0;
+        for (auto successor : llvm::enumerate(pred->getSuccessors())) {
```
will it look nicer with `llvm::find_if`?

pifon2a commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,281 @@
+        auto successorOps = terminator->getSuccessorOperands(successorIndex);
+        // Build the actual mapping of values to their immediate aliases.
+        for (auto arg : block.getArguments()) {
+          auto value = successorOps[arg.getArgNumber()];
```
Please, rename `value` into smth more meaningful to remove the comment above.

pifon2a commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,281 @@
+        }
+        // Get the terminator and the values that will be passed to our block.
+        auto terminator = pred->getTerminator();
+        auto successorOps = terminator->getSuccessorOperands(successorIndex);
```
```
auto successorOps =  pred->getTerminator()->getSuccessorOperands(successorIndex);
```

pifon2a commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h:
```diff
@@ -0,0 +1,185 @@
+  /// Creates a new BufferAssignment analysis that computes liveness of values
+  /// (including their aliases) accross block boundaries to place allocs and
+  /// deallocs.
+  BufferAssignment(Operation* op);
```
explicit?

dfki-ehna commented on tensorflow/compiler/mlir/xla/tests/buffer-assignment.mlir:
```diff
@@ -0,0 +1,131 @@
+// RUN: tf-opt -test-buffer-assignment -split-input-file %s -o - 2>&1 | FileCheck %s -dump-input-on-failure
```
We have to keep 2>&1 in order to dump the std error output stream.

dfki-mako committed:
Addressed reviewer comments.

dfki-mako committed:
Refactored BufferAssignment analysis into two distinct parts:
- BufferAssignmentLegalizer helper class that can be used during legalization.
- BufferAssignmentPass that moves alloc and dealloc nodes (if any) into the
  right positions. If there are not associated dealloc nodes for a given
  alloc node, this pass inserts a DeallocOp in the right place automatically.

dfki-ehna commented:
Refactored BufferAssignment analysis into two distinct parts:
- BufferAssignmentLegalizer helper class that can be used during legalization.
- BufferAssignmentPass that moves alloc and dealloc nodes (if any) into the
  right positions. If there are not associated dealloc nodes for a given
  alloc node, this pass inserts a DeallocOp in the right place automatically.

Commit:
Adding support for function and block signature conversion

sherhut commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,499 @@
+  /// Creates a new positions tuple including alloc and dealloc positions.
+  BufferAssignmentPositions(Operation* allocPosition,
+                            Operation* deallocPosition)
+      : allocPosition(allocPosition), deallocPosition(deallocPosition) {}
```
Should this also use `OpBuilder::InsertPoint`?

sherhut commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,499 @@
+    getFunction().walk([&](AllocOp alloc) {
+      // Find already associated dealloc nodes.
+      auto deallocs = analysis.findAssociatedDeallocs(alloc);
+      assert(deallocs.size() < 2 &
```
Is this `&` intentional?

sherhut commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,499 @@
+/// analysis.
+struct BufferAssignmentTestPass : mlir::FunctionPass<BufferAssignmentTestPass> {
+  void runOnFunction() override {
+    llvm::errs() << "Testing : " << getFunction().getName() << "\n";
```
Leftover from debugging?

dfki-ehna replied:
Not actually. This is the approach we used for liveness analysis testing as well. Furthermore, it's used in the test cases to match the functions.

sherhut replied:
Ah, I see. I got confused by the `llvm::errs()`. Why is this printing to `stderr` and not `stdout`?

dfki-mako replied:
Done

sherhut commented:
Could you provide an example how the use of this would look like for hlo->lhlo? It is not clear to me how this composes.

dfki-ehna commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,499 @@
+/// Computes the actual position to place allocs for the given value.
+BufferInsertPosition BufferAssignmentLegalizer::computeAllocPosition(
+    Value value) const {
+  if (auto arg = value.dyn_cast<BlockArgument>()) {
```
No, for instance, consider the case with a critical edge A->B, B->C, A->C where C has a single block argument. We are going to have two Allocs in A and B and a single Dealloc for block argument in C. One Alloc remains without Dealloc.

sherhut replied:
You could split critical edges. So you would get A->B, B->C, A->Split->C?

dfki-mako replied:
As we have already discussed internally, it is much easier to keep the CFG and to place the alloc operation in the dominator. Furthermore, it simplifies our current BA analysis.

dfki-ehna commented:
The intended use of this buffer assignment is that the dialect-expert creates an instance of BufferAssignmentLegalizer at the beginning of the pass (i.e. in the runOnFunction()) and passes this instance to all conversion-pattern classes using OwningRewritePatternList instance.
These conversion pattern classes could inherit from BufferAssignmentOpConversionPattern that has a protected BufferAssignmentLegalizer member. This will avoid creating BufferAssignmentLegalizer multiple times. In each conversion pattern class that the dialect-expert needs to allocate a buffer, they do the following:

BufferAssignmentOpConversionPattern<HloOpTy>::bufferAssignment->computeAllocPosition(result);
auto alloc = allocPosition.insertAlloc<AllocOp>(result, loc, memref_type); // we are going to remove result from the list of arguments in the next commit since it's unnecessary.

Later on, --buffer-assignment flag must be passed to tf-opt to move Alloc and Dealloc operations in their proper positions. If there is no Dealloc for a buffer, it will be automatically inserted.

dfki-ehna commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h:
```diff
@@ -0,0 +1,140 @@
+/// Prepares a buffer assignment phase. It can place (user-defined) alloc
+/// nodes. This simplifies the integration of the actual buffer-assignment
+/// pass. Sample usage:
+///   BufferAssignmentLegalizer baHelper(value);
```
'value' is the result of an operation that defines 'value' in the source dialect (like %0 = xla_hlo.exp(%arg)). The BufferAssignmentLegalizer can be built once and pass to all conversion pattern. Please take a look at the example.

sherhut replied:
This is not obvious from the comment. I would have expected that it takes the operation whose regions are to be legalized or similar. That is why it was not clear to me what value to pass.

dfki-mako replied:
We fixed the sample :)

dfki-ehna commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h:
```diff
@@ -0,0 +1,140 @@
+/// Note: this class is intended to be used during legalization. In order
+/// to move alloc and dealloc nodes into the right places you can use the
+/// createBufferAssignmentPass() function.
+class BufferAssignmentLegalizer {
```
This was also a part of our discussion about how we should name it. However, placing the Alloc is not the only purpose of this class, and signature conversion is also a part of it. 
We thought we might add a template ReturnOp-Converter method to this class. It replaces a non-void return to a copy operation following with a void return in the target dialect.  If we follow your suggestion to make the signature conversion as a static method, then we can rename it to BufferAllocationPlacer.

sherhut replied:
I think it composes better if the patterns are available separately. To use this, one has to write conversion patterns for the dialect ops and apply them, which would also be the right time to apply the return pattern and signature conversion.

dfki-mako replied:
We have extended our public BA API by exposing signature and return-operation conversion.

dfki-ehna commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h:
```diff
@@ -0,0 +1,140 @@
+  /// TODO(dfki): Value can be removed from arguments.
+  template <typename AllocOpT, typename... Args>
+  AllocOpT insertAlloc(Value value, Args... args) const {
+    OpBuilder allocBuilder(value.getDefiningOp());
```
Yes, you are right. As we mentioned in todo, we wanted to get rid of this argument.

dfki-mako replied:
We have removed the whole class.

Commit:
Providing template non-void-to-void-returnOp conversion pattern for buffer assignment

dfki-mako committed:
Removed BufferInsertPosition and extended our BufferAssignment API by exposing signature and return op conversion.

sherhut reviewed (dismissed):
Let's get this landed and improve further in tree. Thanks!

dfki-mako committed:
Updated BA to support the latest API and IR changes.

gbaned commented:
@dfki-ehna Can you please address Ubuntu Sanity errors? Thanks!

sherhut commented on tensorflow/compiler/mlir/BUILD:
```diff
@@ -74,6 +74,7 @@ cc_library(
         "//tensorflow/compiler/mlir/tensorflow:tf_legalize_hlo",
         "//tensorflow/compiler/mlir/xla:hlo",
         "//tensorflow/compiler/mlir/xla:hlo_legalize_to_lhlo",
+        "//tensorflow/compiler/mlir/xla:buffer_assignment",
```
This needs to be sorted. See the internal CI failure.

dfki-ehna replied:
Done.

sherhut reviewed (approved):
Thanks!

joker-eph commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,500 @@
+      }
+    }
+    return result;
+  }
```
Is there a test covering this function? (our test coverage shows it uncovered)

joker-eph commented on tensorflow/compiler/mlir/xla/transforms/buffer_assignment.cc:
```diff
@@ -0,0 +1,500 @@
+      }
+    });
+  };
+};
```
I'm missing where is this pass covered by tests?

joker-eph commented:
Lacking my comment being addressed, I issued a revert here: https://github.com/tensorflow/tensorflow/pull/38291

byronyi commented:
This unfortunately leaves this PR to a terminal state. @dfki-ehna would you mind to open a new one? Thanks!

dfki-mako commented:
@joker-eph Oh, I see 🔢 Maybe we can add a test dialect including a legalization step in order to simulate lowering from one dialect to another one (or even the same) with buffers?

joker-eph commented:
Why do we need a test dialect? I may be missing something but the description of the pass is:

```
 /// The actual buffer assignment pass that moves alloc and dealloc nodes into
 /// the right positions. 
```

I was expecting this to be able to run on something like LHLO?

pifon2a commented:
@joker-eph The test dialect will be needed if we want to have this pass in MLIR core, since there is no  LHLO there. I think @dfki-mako was assuming this.

dfki-mako commented:
@joker-eph I have already thought about this transfer and future features at the same time. It might be useful to have a small test dialect that contains additional features that we might want to think about and include in the general BA pass.

dfki-mako commented:
@joker-eph Please note that this is not a strict requirement at the moment. We can include several Linalg-based tests in the MLIR core version and apply BA to them.

joker-eph commented:
I am a bit lost (but it is quite late here, my brain may be slow...): I was looking at this PR in TensorFlow, in a disconnected way from Core, and I see code that does not seem to have any test here.
Regardless of plans about upstreaming, I expect that the development in MLIR-TensorFlow follows roughly the same set of good practices as upstream, starting with complete coverage of new code with lit tests. Unless I'm mistaken there is a non-tested pass in this PR?

dfki-mako commented:
Yes and no. We had to redesign several parts and split some features (regarding legalization and use) into other PRs. No, because the test pass basically computes all required information to move allocs and deallocs. Yes, because the actual movement part is not covered.

dfki-ehna commented:
@joker-eph you are right with respect to this PR. It has never been our intention to push an untested pass. We address the missing-tests issue and add a complete test case to test the whole BA pass including movement of allocs and deallocs.

joker-eph commented:
>  We address the missing-tests issue and add a complete test case to test the whole BA pass including movement of allocs and deallocs.

Is there already a PR up for review?
